### PR TITLE
[1/2][Metrics] Support the logical resource metrics 

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This PR adds the logical resource report. I think we could also report this from GCS, but raylet is closer to the source of truth, so I added there. 

Supporting placement_group_reserved_available,  placement_group_reserved_used states will be in a follow up PR. Right now, we just ignore reporting them.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by rkooo567** [comment]
Ping on review!

File: src/ray/stats/metric_defs.cc [hunk 1]
```diff
@@ -169,6 +169,14 @@ DEFINE_stats(scheduler_failed_worker_startup_total,
              (),
              ray::stats::GAUGE);
 
+/// Raylet Resource Manager
+DEFINE_stats(resources,
+             // TODO(sang): Support placement_group_reserved_available | used
+             "Logical Ray resources broken per state {AVAILABLE, USED}",
+             ("Name", "State"),
+             (),
+             ray::stats::GAUGE);
```
Comment (ericl) [inline src/ray/stats/metric_defs.cc:178]: Feel free to punt this to later, but it would be nice to add a IN_PLACEMENT_GROUP breakdown as well.

File: src/ray/common/bundle_spec.cc [hunk 2]
```diff
@@ -144,6 +144,33 @@ std::string GetOriginalResourceNameFromWildcardResource(const std::string &resou
   return match_groups[1].str();
 }
 
+std::optional<PgFormattedResourceData> ParsePgFormattedResource(
```
[the remark this answers is not in the document]
Comment (rkooo567) [reply src/ray/common/bundle_spec.cc:147]: let me take out the parsing code as a helper func

File: src/ray/stats/metric_defs.cc (hunk 1 above)
Comment (rkooo567) [reply src/ray/stats/metric_defs.cc:178]: that will be done in a follow up PR!

File: src/ray/raylet/scheduling/local_resource_manager_test.cc [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    ASSERT_TRUE(resource_usage_map.find(node_ip_resource) == resource_usage_map.end());
+    // Verify pg resources are not reported.
+    ASSERT_TRUE(resource_usage_map.find(pg_wildcard_resource) ==
+                resource_usage_map.end());
+    ASSERT_TRUE(resource_usage_map.find(pg_index_0_resource) == resource_usage_map.end());
+    ASSERT_TRUE(resource_usage_map.find(pg_index_1_resource) == resource_usage_map.end());
+  }
+}
+
+}  // namespace ray
```
Comment (ericl) [inline src/ray/raylet/scheduling/local_resource_manager_test.cc:125]: Great tests

File: src/ray/common/bundle_spec.cc [hunk 4]
```diff
@@ -118,30 +118,57 @@ std::string FormatPlacementGroupResource(const std::string &original_resource_na
...
 }
 
-bool IsBundleIndex(const std::string &resource,
-                   const PlacementGroupID &group_id,
-                   const int bundle_index) {
-  return resource.find(kGroupKeyword + std::to_string(bundle_index) + "_" +
-                       group_id.Hex()) != std::string::npos;
-}
-
 std::string GetOriginalResourceName(const std::string &resource) {
```
Comment (ericl) [inline src/ray/common/bundle_spec.cc:121]: Can we split the PG parsing refactoring into a separate PR? This is also a bit of a scary change that would be nice to avoid.

**Review by ericl — COMMENTED** [review]
The non-PG parts LGTM.

File: src/ray/common/bundle_spec.cc (hunk 4 above)
Comment (rkooo567) [reply src/ray/common/bundle_spec.cc:121]: Would you like to merge this after 2.1 release? Btw, I'd like to keep `ParsePgFormattedResource` to avoid reporting pg resources to metrics if possible!

**Comment by rkooo567** [comment]
Will have 2 follow up PRs;

1. pg parsing refactoring
2. support used-by-pg tags

File: src/ray/common/bundle_spec.cc [hunk 5]
```diff
@@ -144,6 +144,40 @@ std::string GetOriginalResourceNameFromWildcardResource(const std::string &resou
   return match_groups[1].str();
 }
 
+std::optional<PgFormattedResourceData> ParsePgFormattedResource(
+    const std::string &resource, bool for_wildcard_resource, bool for_indexed_resource) {
```
Comment (ericl) [inline src/ray/common/bundle_spec.cc:148]: Could we use the above API calls instead of adding a new one?

**Review by ericl — COMMENTED** [review]
Shall we remove all the pg formatting code as well?

File: src/ray/common/bundle_spec.cc (hunk 5 above)
Comment (rkooo567) [reply src/ray/common/bundle_spec.cc:148]: The existing API cannot handle non-wildcard resources. There's `GetOriginalResourceName`, but this API assumes we already know the resources are pg resources. I can change this in this PR and fix, but I feel like the risk is pretty low...

**Review by ericl — APPROVED** [review]
